File: src/com/pratham/DoublyLL.java
package com.pratham;

public class DoublyLL {
    static Node head;

    static class Node{
        int data;
        Node next;
        Node prev;

        Node(int val){
            data = val;
            next = null;
        }
    }
    public static void main(String[] args) {
        DoublyLL llist = new DoublyLL();

        // Creating 5 Nodes
        Node first = new Node(10);
        head = first;

        Node second = new Node(20);
        first.next = second;
        second.prev = first;

        Node third = new Node(30);
        second.next = third;
        third.prev = second;

        Node fourth = new Node(40);
        third.next = fourth;
        fourth.prev = third;

        Node fifth = new Node(50);
        fourth.next = fifth;
        fifth.prev = fourth;

        System.out.println("LinkedList Consist of 5 Nodes: ");
        llist.display();

        llist.insertBegin(369);
        llist.insertEnd(370);
        llist.insertSpecific(20,100);

        System.out.println("After insertion at beginning,at end,at specific position: ");
        llist.display();
    }

    void display(){
        Node ptr;
        ptr = head;
        while (ptr!=null){
            System.out.print(ptr.data + " ");
            ptr=ptr.next;
        }
        System.out.println();
    }

    void insertBegin(int val){
        Node newNode = new Node(val);
        newNode.next = head;
        head = newNode;
    }

    void insertEnd(int val){
        Node newNode = new Node(val);
        Node ptr = head;
        while (ptr.next!=null){
            ptr = ptr.next;
        }
        newNode.prev = ptr;
        ptr.next = newNode;
    }

    void insertSpecific(int val,int data){
        Node newNode = new Node(data);
        Node ptr = head;
        while (ptr.data!=val){
            ptr = ptr.next;
        }
        newNode.next = ptr.next;
        ptr.next = newNode;
        newNode.prev = ptr;
    }

}

File: C++ Implementation/doublyLL.cpp
#include<iostream>
using namespace std;
class Node{
    int data;
    Node *head;
    Node *next,*prev;

    public:
        Node(){
            head = NULL;
            prev=next=NULL;
        }

        void insertBegin(int data){
            if(head==NULL){
                Node* newNode = new Node();
                newNode->data = data;
                head = newNode;
            } else{
                Node* newNode = new Node();
                newNode->data = data;
                Node* ptr = head;
                newNode->next = ptr->prev;
                ptr->prev = newNode;
                head = newNode;
            }
        }

        void insertEnd(int data){
            if(head==NULL){
                Node* newNode = new Node();
                newNode->data = data;
                head = newNode;
            } else{
                Node* newNode = new Node();
                newNode->data = data;
                newNode->next = NULL;
                Node *ptr = head;
                while (ptr->next!=NULL){
                    ptr = ptr->next;
                }
                ptr->next = newNode;
                newNode->prev = ptr;
            }
        }

        void insertBefore(int ele, int data){
            Node* newNode = new Node();
            newNode->data = data;
            Node* ptr = head;
            Node* temp = head;
            while (ptr->data!=ele){
                temp = ptr;
                ptr = ptr->next;
            }
            newNode->next = temp->next;
            ptr->prev = newNode;
            temp->next = newNode;
            newNode->prev = temp;
        }

        void deleteBegin(){
            Node* ptr = head;
            Node* temp = head;
            ptr = ptr->next;
            ptr->prev = NULL;
            head = ptr;
            delete temp;
        }

        void deleteEnd(){
            Node* ptr = head;
            while (ptr->next!=NULL){
                ptr = ptr->next;
            }
            Node* temp = ptr->prev;
            temp->next = NULL;
            delete ptr;
        }

        void deleteBefore(int ele){
            Node* ptr = head;
            while(ptr->data!=ele){
                ptr=ptr->next;
            }
            Node* temp = ptr->prev;
            Node* helper = temp->prev;
            helper->next = ptr;
            ptr->prev = helper;
            delete temp;
        }

        void display(){
            Node* ptr = head;
            while (ptr!=NULL){
                cout << ptr->data << " ";
                ptr = ptr->next;
            }
            cout << endl;
        }
};
int main()
{
    Node ll;
    ll.insertBegin(10);
    ll.insertEnd(20);
    ll.insertEnd(30);
    ll.insertEnd(50);
    ll.insertBefore(50,40);
    ll.display();

    ll.deleteBegin();
    ll.display();
    ll.deleteEnd();
    ll.display();
    ll.deleteBefore(40);
    ll.display();

    

    return 0;
}